File: src/android/app/src/main/java/com/selkacraft/alice/comm/network/SyncProtocol.kt
package com.selkacraft.alice.comm.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Network sync protocol message types and serialization.
 * Matches the desktop C++ implementation for bidirectional sync.
 */

private val json = Json {
    ignoreUnknownKeys = true
    isLenient = true
    encodeDefaults = true
}

@Serializable
data class SyncMessage(
    val type: String,
    val timestamp: Long = System.currentTimeMillis(),
    val sender: String = "android",
    val payload: Map<String, kotlinx.serialization.json.JsonElement> = emptyMap()
) {
    fun serialize(): String = json.encodeToString(this)

    companion object {
        fun deserialize(data: String): SyncMessage? {
            return try {
                json.decodeFromString<SyncMessage>(data)
            } catch (e: Exception) {
                null
            }
        }

        fun stateUpdate(
            motorPosition: Int,
            depth: Float,
            confidence: Float,
            focusMode: String,
            enabled: Boolean,
            activelyFocusing: Boolean,
            facesDetected: Int,
            motorConnected: Boolean,
            realSenseConnected: Boolean
        ): SyncMessage {
            val payloadMap = buildJsonMap {
                put("motorPosition", motorPosition)
                put("depth", depth)
                put("depthConfidence", confidence)
                put("focusMode", focusMode)
                put("isEnabled", enabled)
                put("isActivelyFocusing", activelyFocusing)
                put("facesDetected", facesDetected)
                put("connectionStates", buildJsonMap {
                    put("motor", if (motorConnected) "Active" else "Disconnected")
                    put("realSense", if (realSenseConnected) "Active" else "Disconnected")
                })
            }
            return SyncMessage(type = "STATE_UPDATE", payload = payloadMap)
        }

        fun motorCommand(position: Int, source: String = "manual", highPriority: Boolean = false): SyncMessage {
            val payloadMap = buildJsonMap {
                put("position", position)
                put("source", source)
                put("priority", if (highPriority) "HIGH" else "NORMAL")
            }
            return SyncMessage(type = "MOTOR_COMMAND", payload = payloadMap)
        }

        fun modeChange(
            mode: String,
            enabled: Boolean,
            confidenceThreshold: Float,
            smoothing: Boolean,
            responseSpeed: Int
        ): SyncMessage {
            val payloadMap = buildJsonMap {
                put("mode", mode)
                put("enabled", enabled)
                put("settings", buildJsonMap {
                    put("confidenceThreshold", confidenceThreshold)
                    put("smoothing", smoothing)
                    put("responseSpeed", responseSpeed)
                })
            }
            return SyncMessage(type = "MODE_CHANGE", payload = payloadMap)
        }

        fun authenticate(token: String): SyncMessage {
            val payloadMap = buildJsonMap {
                put("token", token)
            }
            return SyncMessage(type = "AUTHENTICATE", payload = payloadMap)
        }

        fun calibrationSync(
            action: String,
            mappingJson: kotlinx.serialization.json.JsonElement? = null
        ): SyncMessage {
            val payloadMap = buildJsonMap {
                put("action", action)
                if (mappingJson != null) {
                    putElement("mapping", mappingJson)
                }
            }
            return SyncMessage(type = "CALIBRATION_SYNC", payload = payloadMap)
        }

        fun streamControl(color: Boolean, depth: Boolean, capture: Boolean): SyncMessage {
            val payloadMap = buildJsonMap {
                put("color", color)
                put("depth", depth)
                put("capture", capture)
            }
            return SyncMessage(type = "STREAM_CONTROL", payload = payloadMap)
        }

        fun heartbeat(uptime: Long, deviceInfo: String = ""): SyncMessage {
            val payloadMap = buildJsonMap {
                put("uptime", uptime)
                put("deviceInfo", deviceInfo)
            }
            return SyncMessage(type = "HEARTBEAT", payload = payloadMap)
        }

        fun measurePosition(x: Float, y: Float): SyncMessage {
            val payloadMap = buildJsonMap {
                put("x", x)
                put("y", y)
            }
            return SyncMessage(type = "MEASURE_POSITION", payload = payloadMap)
        }
    }
}

/**
 * Helper to build a JSON map for payload construction.
 */
fun buildJsonMap(
    builder: JsonMapBuilder.() -> Unit
): Map<String, kotlinx.serialization.json.JsonElement> {
    return JsonMapBuilder().apply(builder).build()
}

class JsonMapBuilder {
    private val map = mutableMapOf<String, kotlinx.serialization.json.JsonElement>()

    fun put(key: String, value: String) {
        map[key] = kotlinx.serialization.json.JsonPrimitive(value)
    }
    fun put(key: String, value: Int) {
        map[key] = kotlinx.serialization.json.JsonPrimitive(value)
    }
    fun put(key: String, value: Long) {
        map[key] = kotlinx.serialization.json.JsonPrimitive(value)
    }
    fun put(key: String, value: Float) {
        map[key] = kotlinx.serialization.json.JsonPrimitive(value)
    }
    fun put(key: String, value: Boolean) {
        map[key] = kotlinx.serialization.json.JsonPrimitive(value)
    }
    fun put(key: String, value: Map<String, kotlinx.serialization.json.JsonElement>) {
        map[key] = kotlinx.serialization.json.JsonObject(value)
    }
    fun putElement(key: String, value: kotlinx.serialization.json.JsonElement) {
        map[key] = value
    }

    fun build() = map.toMap()
}

object SyncConstants {
    const val DEFAULT_PORT = 8765
    const val STATE_UPDATE_INTERVAL_MS = 100L // 10 Hz
    const val HEARTBEAT_INTERVAL_MS = 5000L
    const val CONNECTION_TIMEOUT_MS = 10000L
}

File: src/desktop/src/network/SyncProtocol.cpp
#include "network/SyncProtocol.h"
#include <QDateTime>

namespace alice {

namespace {
    QString typeToString(SyncMessageType type) {
        switch (type) {
        case SyncMessageType::StateUpdate:     return "STATE_UPDATE";
        case SyncMessageType::MotorCommand:    return "MOTOR_COMMAND";
        case SyncMessageType::ModeChange:      return "MODE_CHANGE";
        case SyncMessageType::CalibrationSync: return "CALIBRATION_SYNC";
        case SyncMessageType::Heartbeat:       return "HEARTBEAT";
        case SyncMessageType::Authenticate:    return "AUTHENTICATE";
        case SyncMessageType::StreamControl:   return "STREAM_CONTROL";
        }
        return "UNKNOWN";
    }

    SyncMessageType stringToType(const QString &str) {
        if (str == "STATE_UPDATE")     return SyncMessageType::StateUpdate;
        if (str == "MOTOR_COMMAND")    return SyncMessageType::MotorCommand;
        if (str == "MODE_CHANGE")      return SyncMessageType::ModeChange;
        if (str == "CALIBRATION_SYNC") return SyncMessageType::CalibrationSync;
        if (str == "HEARTBEAT")        return SyncMessageType::Heartbeat;
        if (str == "AUTHENTICATE")     return SyncMessageType::Authenticate;
        if (str == "STREAM_CONTROL")   return SyncMessageType::StreamControl;
        return SyncMessageType::Heartbeat;
    }
}

QByteArray SyncMessage::serialize() const {
    QJsonObject obj;
    obj["type"] = typeToString(type);
    obj["timestamp"] = timestamp > 0 ? timestamp : QDateTime::currentMSecsSinceEpoch();
    obj["sender"] = sender;
    obj["payload"] = payload;
    return QJsonDocument(obj).toJson(QJsonDocument::Compact);
}

std::optional<SyncMessage> SyncMessage::deserialize(const QByteArray &data) {
    QJsonParseError err;
    auto doc = QJsonDocument::fromJson(data, &err);
    if (err.error != QJsonParseError::NoError || !doc.isObject()) return std::nullopt;

    auto obj = doc.object();
    SyncMessage msg;
    msg.type = stringToType(obj["type"].toString());
    msg.timestamp = obj["timestamp"].toVariant().toLongLong();
    msg.sender = obj["sender"].toString();
    msg.payload = obj["payload"].toObject();
    return msg;
}

SyncMessage SyncMessage::stateUpdate(int motorPosition, float depth, float confidence,
                                      const QString &focusMode, bool enabled,
                                      bool activelyFocusing, int facesDetected,
                                      bool motorConnected, bool realSenseConnected) {
    SyncMessage msg;
    msg.type = SyncMessageType::StateUpdate;
    msg.sender = "desktop";
    msg.payload = QJsonObject{
        {"motorPosition", motorPosition},
        {"depth", static_cast<double>(depth)},
        {"depthConfidence", static_cast<double>(confidence)},
        {"focusMode", focusMode},
        {"isEnabled", enabled},
        {"isActivelyFocusing", activelyFocusing},
        {"facesDetected", facesDetected},
        {"connectionStates", QJsonObject{
            {"motor", motorConnected ? "Active" : "Disconnected"},
            {"realSense", realSenseConnected ? "Active" : "Disconnected"}
        }}
    };
    return msg;
}

SyncMessage SyncMessage::motorCommand(int position, const QString &source, bool highPriority) {
    SyncMessage msg;
    msg.type = SyncMessageType::MotorCommand;
    msg.payload = QJsonObject{
        {"position", position},
        {"source", source},
        {"priority", highPriority ? "HIGH" : "NORMAL"}
    };
    return msg;
}

SyncMessage SyncMessage::modeChange(const QString &mode, bool enabled,
                                     float confidenceThreshold, bool smoothing, int responseSpeed) {
    SyncMessage msg;
    msg.type = SyncMessageType::ModeChange;
    msg.payload = QJsonObject{
        {"mode", mode},
        {"enabled", enabled},
        {"settings", QJsonObject{
            {"confidenceThreshold", static_cast<double>(confidenceThreshold)},
            {"smoothing", smoothing},
            {"responseSpeed", responseSpeed}
        }}
    };
    return msg;
}

SyncMessage SyncMessage::calibrationSync(const QString &action, const QJsonObject &mappingJson) {
    SyncMessage msg;
    msg.type = SyncMessageType::CalibrationSync;
    msg.payload = QJsonObject{
        {"action", action},
        {"mapping", mappingJson}
    };
    return msg;
}

SyncMessage SyncMessage::heartbeat(int64_t uptime, const QString &deviceInfo) {
    SyncMessage msg;
    msg.type = SyncMessageType::Heartbeat;
    msg.payload = QJsonObject{
        {"uptime", static_cast<qint64>(uptime)},
        {"deviceInfo", deviceInfo}
    };
    return msg;
}

SyncMessage SyncMessage::authenticate(const QString &token) {
    SyncMessage msg;
    msg.type = SyncMessageType::Authenticate;
    msg.payload = QJsonObject{{"token", token}};
    return msg;
}

SyncMessage SyncMessage::streamControl(bool color, bool depth, bool capture) {
    SyncMessage msg;
    msg.type = SyncMessageType::StreamControl;
    msg.payload = QJsonObject{
        {"color", color},
        {"depth", depth},
        {"capture", capture}
    };
    return msg;
}

} // namespace alice
